Feat/bitcoin flag newer format

criadoperez opened the pull request:
Adding Bitcoin node feature to the new format

criadoperez commented:
@wileyj please review whenever you can.
No user config files (yaml, toml or config) are ever modified during runtime for this feature.

wileyj commented on compose-files/extra-services/bitcoin.yaml:
```diff
@@ -0,0 +1,14 @@
+version: "3.8"
```
why version 3.8? is there something present in 3.8 vs 3?

criadoperez replied:
I'm not looking for the oldest version that works, but a current stable version that works and will be compatible with future versions.
3.8 seems like a good balance for me. But if you prefer 3, I'm sure it will also work fine also.

wileyj replied:
I have no problem with using 3.8 per se, but if it's used for this compose file it should also be used everywhere. typically you only use compose point releases is if there is a specific syntax/feature you're trying to use. else, it's more common to use `2` or `3`. in this repo, it was setup for version 3, since the point release didn't have a compelling use (at the same time, `3` is compatible with older version of docker). 
https://docs.docker.com/compose/compose-file/compose-versioning/#version-38

that said, if you find some interesting features available that outweigh the benefit of working with older docker versions, `3.8` is perfectly fine - just be sure to update all of the compose files.

wileyj replied:
additionally, version 2 of compose still *works* fine - you're just not going to have access to the features in version 3 of the format.

criadoperez replied:
I completely agree that we must be coherent and all files should use the same version number. I was going to switch it to version 3, but reviewing the features per version, I see that [version 3.5](https://docs.docker.com/compose/compose-file/compose-versioning/#version-35) introduced the `name for networks, secrets and configs`, which funny enough we are using both in my new files and in your old file that use version 3.
Version 3.5 requires Docker Engine version 17.09
So, I'm changing all files to version 3.5 and adding the Docker Engine minimum version to the requirements docs.
If for whatever reason you still prefer v.3 we can always switch them all again, but at least they will always all have the same version.

wileyj commented on conf/testnet/bitcoin.conf:
```diff
@@ -0,0 +1,4 @@
+rpcuser=stacks
```
this file isn't needed, only the sample file. if this file doesn't exist locally, the script should copy sample to this one (and added to gitignore)

criadoperez replied:
That is correct. I must have uploaded it without realizing before adding it to gitignore. (it's already in gitignore). I'll delete the file in a new commit.

wileyj commented:
The Dockerfiles moved out of `./Dockerfiles` dir to `./compose-files` should be moved back - they're specifically meant to be separate since they're **not** docker compose files and can be built separately from docker-compose.

wileyj commented on sample.env:
```diff
@@ -51,3 +55,9 @@ STACKS_BLOCKCHAIN_API_VERSION=3.0.3
+###############################
+## Bitcoin config
+##
+BITCOIN_BLOCKCHAIN_FOLDER=${SCRIPTPATH}/persistent-data/blockchain-bitcoin
```
using this as-is will not work if you run testnet, then restart with mainnet

criadoperez replied:
> using this as-is will not work if you run testnet, then restart with mainnet

Why not? The testnet data and the mainnet data are both stored in different subfolders. I'll test it anyway to see if I encounter the problem you mention.

wileyj replied:
mainly, it adds extra complexity vs using the current format of `persistent-data/$NETWORK/<data>`. moving bitcoin data to a new folder would require some code changes (particularly for the `reset` option which deletes persistent data.

criadoperez replied:
ok. I removed ${SCRIPTPATH} and verified it still works fine.

wileyj commented on conf/mainnet/Config-with-bitcoin-flag.toml.sample:
```diff
@@ -0,0 +1,19 @@
+retry_count = 255
+events_keys = ["*"]
+
+[burnchain]
```
this model isn't ideal since it requires hardcoding values that could change

criadoperez replied:
Let's talk about this in a call.
It's a sample file with defaults that the user can change if he wishes but the script will never modify it.

wileyj commented:
functions `setup_bitcoin_up` and `setup_bitcoin_down` are identical. is that by design?

wileyj commented on manage.sh:
```diff
@@ -657,10 +682,12 @@ docker_up() {
-	fi
+	[[ ! -f "${SCRIPTPATH}/conf/${NETWORK}/Config-with-bitcoin-flag.toml" ]] && cp "${SCRIPTPATH}/conf/${NETWORK}/Config-with-bitcoin-flag.toml.sample" "${SCRIPTPATH}/conf/${NETWORK}/Config-with-bitcoin-flag.toml"
+	[[ ! -f "${SCRIPTPATH}/conf/${NETWORK}/bitcoin.conf" ]] && cp "${SCRIPTPATH}/conf/${NETWORK}/bitcoin.conf.sample" "${SCRIPTPATH}/conf/${NETWORK}/bitcoin.conf"
+#	if [[ "${NETWORK}" == "private-testnet" ]]; then
```
incorrect indenting for comments

criadoperez replied:
fixed

wileyj commented on compose-files/common.yaml:
```diff
@@ -33,7 +33,7 @@ services:
         "start",
         "--config",
-        "/src/stacks-blockchain/Config.toml",
+        "${CONFIG_TOML_TO_USE:-/src/stacks-blockchain/Config.toml}",
```
This will always be set to `/src/stacks-blockchain/Config.toml` *unless* manual changes are made to the .env file as-is. It's better to assign a variable and export it in the script vs relying on manual changes.

criadoperez replied:
fixed. Thanks.

criadoperez commented on compose-files/extra-services/bitcoin/Dockerfile:
```diff
@@ -1,12 +1,11 @@
 # COPY ./docker-entrypoint.sh /docker-entrypoint.sh
 # RUN chmod 755 /docker-entrypoint.sh /make_binary_dist.sh && mkdir -p /root/.bitcoin
-RUN chmod 755 /make_binary_dist.sh && mkdir -p /root/.bitcoin
+RUN chmod 755 /make_binary_dist.sh && mkdir -p /root/.bitcoin && mkdir -p /etc/bitcoin
```
same applies to /root/.bitcoin that is also mounted at runtime. Both are redundant.
I added it to be consistent with the current code logic.
I would either leave it as it is or delete both. I'll delete them both on a new commit if you prefer (unless you tell me otherwise)

wileyj replied:
makes sense - remove them both

criadoperez replied:
ok. removed `&& mkdir -p /etc/bitcoin`

criadoperez commented:
> The Dockerfiles moved out of `./Dockerfiles` dir to `./compose-files` should be moved back - they're specifically meant to be separate since they're **not** docker compose files and can be built separately from docker-compose.

Yes and no. 

I was kind of expecting this point to be an issue... :-)

Because the dockerfile has some variables it can never run by itself directly from docker alone. They only work if called with docker-compose from the script itself. When I have dockerfiles that are dependent of the dockercompose I prefer to have them inside the dockercompose folder. 

However, I understand this is just a personal preference and both options work fine.

As you prefer the other logic, I'll move it back to the Dockerfiles folder in the next commit, no problem.

criadoperez commented:
Coverted to draft until all requested changes are completed. 
I will remove draft status once they are all resolved.

wileyj commented:
> > The Dockerfiles moved out of `./Dockerfiles` dir to `./compose-files` should be moved back - they're specifically meant to be separate since they're **not** docker compose files and can be built separately from docker-compose.
> 
> Yes and no.
> 
> I was kind of expecting this point to be an issue... :-)
> 
> Because the dockerfile has some variables it can never run by itself directly from docker alone. They only work if called with docker-compose from the script itself. When I have dockerfiles that are dependent of the dockercompose I prefer to have them inside the dockercompose folder.
> 
> However, I understand this is just a personal preference and both options work fine.
> 
> As you prefer the other logic, I'll move it back to the Dockerfiles folder in the next commit, no problem.

Not quite  - the only variable used is `BTC_VERSION` which is a build arg (default is set in the file, but can be overridden with a build arg of a different version. 
full disclosure - when i build local bitcoin docker images, this is the dockerfile i use so it *does* build correctly outside of compose

wileyj commented:
> > The Dockerfiles moved out of `./Dockerfiles` dir to `./compose-files` should be moved back - they're specifically meant to be separate since they're **not** docker compose files and can be built separately from docker-compose.
> 
> Yes and no.
> 
> I was kind of expecting this point to be an issue... :-)
> 
> Because the dockerfile has some variables it can never run by itself directly from docker alone. They only work if called with docker-compose from the script itself. When I have dockerfiles that are dependent of the dockercompose I prefer to have them inside the dockercompose folder.
> 
> However, I understand this is just a personal preference and both options work fine.
> 
> As you prefer the other logic, I'll move it back to the Dockerfiles folder in the next commit, no problem.

Thanks - the overall idea of the last big merge i made was to simplify the directories. main goal was to have all the compose-files (and only the compose files) in this dir.

criadoperez commented on sample.env:
```diff
@@ -37,6 +37,10 @@ STACKS_LOG_DEBUG=0
 ##    ** if you kill this thread before it's done, it can cause a corrupt chainstate **
 STACKS_SHUTDOWN_TIMEOUT=1200
+#
+CONFIG_TOML_WITHOUT_BITCOIN_FLAG=/src/stacks-blockchain/Config.toml
```
Sorry, but I'm not sure I understand your point here.
The default value of the variable is set in the `.env` file so its not hardcoded in the script.
You prefer to hardcode it in the script instead?

wileyj replied:
hardcoded, no - it should be *created* in the script. the value that the container uses should not be changed at runtime, it should be changed via the script so the container always uses the desired file. 

it's easier to detect/catch errors in the script vs trying to determine if an incorrect file is bind mounted

criadoperez replied:
Yes, you are right. Ok, moved them to `manage.sh`.

wileyj commented on README.md:
````diff
@@ -122,19 +126,25 @@ _You may also use a symlink as an alternative to copying: `ln -s sample.env .env
 
 ```bash
-./manage.sh -n <network> -a start
+./manage.sh -n <network> -a up
````
technically, this is still correct - but using the sysv standard `start` is more accurate

criadoperez replied:
Yes, technically they both work and in my opinion `start` and `stop` is more intuitive than `up` and `down`. 
However, we need to be coherent with one term or another. I find it very confusing for a user that reads `start` on the docs and on the command line you see `up`.
```bash
$ ./manage.sh 

Usage:                    
    ./manage.sh -n <network> -a <action> <optional args> 
        -n|--network: [ mainnet | testnet | mocknet | bns ] 
        -a|--action: [ up | down | logs | reset | upgrade | import | export | bns ] 
    optional args:        
        -f|--flags: [ proxy,bitcoin ] 
        export: combined with 'logs' action, exports logs to a text file 
    ex: ./manage.sh -n mainnet -a up -f proxy,bitcoin 
    ex: ./manage.sh --network mainnet --action up --flags proxy 
    ex: ./manage.sh -n mainnet -a logs export 
```

I would recommend changing it to one coherent term (any term). We change everything (including the command line output) to `start` and `stop`.?

wileyj commented on README.md:
````diff
@@ -122,19 +126,25 @@ _You may also use a symlink as an alternative to copying: `ln -s sample.env .env
 
 ```bash
-./manage.sh -n <network> -a stop
+./manage.sh -n <network> -a down
````
technically, this is still correct - but using the sysv standard `stop` is more accurate

criadoperez replied:
explained on the comment above. We can continue the conversation there.

wileyj commented on README.md:
````diff
@@ -353,9 +379,9 @@ _**Database Issues**_:
 
 ```bash
-./manage.sh -n <network> -a stop
+./manage.sh -n <network> -a down
````
see above re: stop

criadoperez replied:
explained above.

wileyj commented on README.md:
```diff
@@ -353,9 +379,9 @@ _**Database Issues**_:
+./manage.sh -n <network> -a down
 ./manage.sh -n <network> -a reset
-./manage.sh -n <network> -a start
+./manage.sh -n <network> -a up
```
see above re: start

criadoperez replied:
explained above.

wileyj commented on README.md:
````diff
@@ -368,7 +394,7 @@ _**API Missing Parent Block Error**_:
 
 ```bash
-./manage.sh -n <network> -a stop
+./manage.sh -n <network> -a down
````
see above re: stop

criadoperez replied:
explained above.

wileyj commented on README.md:
````diff
@@ -377,7 +403,7 @@ _**API Missing Parent Block Error**_:
 
 ```bash
-./manage.sh -n <network> -a stop
+./manage.sh -n <network> -a down
````
see above re: stop

criadoperez replied:
explained above.

wileyj commented on .gitignore:
```diff
@@ -28,11 +27,10 @@ Session.vim
 node_modules/
 stacks-node-follower/Config.toml
-configurations/*/Config.toml
-configurations/private-testnet/*.toml
```
I wouldn't address private-testnet files in this PR, it's out of scope

criadoperez replied:
ok. added them back in again.

wileyj commented on compose-files/common.yaml:
```diff
@@ -1,4 +1,4 @@
-version: "3"
+version: "3.5"
```
what features are we using from 3.5 vs 3?

criadoperez replied:
I quote a message from above:

> I completely agree that we must be coherent and all files should use the same version number. I was going to switch it to version 3, but reviewing the features per version, I see that [version 3.5](https://docs.docker.com/compose/compose-file/compose-versioning/#version-35) introduced the `name for networks, secrets and configs`, which funny enough we are using both in my new files and in your old file that use version 3. Version 3.5 requires Docker Engine version 17.09 So, I'm changing all files to version 3.5 and adding the Docker Engine minimum version to the requirements docs. If for whatever reason you still prefer v.3 we can always switch them all again, but at least they will always all have the same version.

criadoperez commented on README.md:
````diff
@@ -105,7 +109,7 @@ Directories will be created on first start that will store persistent data under
 ```
 
-2. **Create/Copy `.env` file**
+2. **Optionally, create/copy `.env` file. If file `.env` doesn't exist when launched it will be created from `sample.env` automatically.**
````
fixed.

criadoperez commented on README.md:
````diff
@@ -143,7 +153,7 @@ _You may also use a symlink as an alternative to copying: `ln -s sample.env .env
 ```
 
-- Export docker logs to `./exported-logs`:
+- Export docker logs to files in folder `./exported-logs`:
````
I though it clarified that `./exported-logs` isn't a file.
What do you think of the following instead:
```
Export docker log files to `./exported-logs`
```
Let me know what you prefer, and I'll change it to the one you like best.

criadoperez commented on README.md:
```diff
@@ -210,6 +220,11 @@ You can run easily run your Stacks node with your own Bitcoin node by adding the
 
+### Bitcoin node configuration
+
+In the `.env` file there is the variable `BITCOIN_BLOCKCHAIN_FOLDER`.
```
No, not yet. 
This can be done for a next version. 
If you are "sharing" the data with another app you normally wouldn't want this app to delete it. But options are always good and it would be a nice to have feature.

criadoperez replied:
Update on this:
I added the reset functionality in a new commit.

criadoperez commented on README.md:
```diff
@@ -64,9 +61,21 @@ sudo curl -L https://github.com/docker/compose/releases/download/${VERSION}/dock
+
+The Docker daemon always runs as the root user so by default you will need root privileges to interact with it.
+
+The script `manage.sh` uses docker, so to avoid the requirement of needing to run the script with root privileges it is prefered to be able to *manage Docker as a non-root user*, following [these simple tests](https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user).
```
Should we add that note or remove the explanation entirely?

I think users would prefer to know that doing this the can run the app without sudo privileges for normal operations (anything except removing data). Much better that needing sudo every time.

Also, there is a confidence issue here. Downloading a script from the Internet that requires sudo everytime you execute it.... is risky and might scare users away.

I added a small explanation with your point, but if you still don't like it, let me know and I'll delete the section entirely.

criadoperez commented:
@wileyj, all your code reviews have been addressed. Thanks for your input.

criadoperez commented:
Added functionality so bitcoin-core logs are also exported with `./manage.sh -n <network> -a logs export` when bitcoin node is running.
I noticed the nginx wasn't exporting the logs neither. Any reason why? I added it also, but if there was a reason to exclude it let me know.

criadoperez commented:
New options when data is reset: 
```
$ sudo ./manage.sh -n testnet -a reset
[sudo] password for alejandro: 
Please confirm what persistent data you wish to delete: 

0. Cancel                 
1. Delete Persistent data for Stacks testnet only and leave Bitcoin blockchain data unaffected. 
2. Delete Persistent data for Stacks testnet and Bitcoin blockchain data in ./persistent-data/blockchain-bitcoin 
3. Delete Persistent data for Bitcoin blockchain data in ./persistent-data/blockchain-bitcoin only. 
Please note that BNS data will never get deleted. 

Type 0, 1, 2 or 3: 2

Ok. Delete Stacks and Bitcoin data. 
[ Success ]      Persistent data deleted for Bitcoin blockchain.
[ Success ]      Persistent data deleted for Stacks testnet.
```

wileyj reviewed (approved):
shipit

criadoperez commented:
> shipit

Cool! I'll resolve the conflicts first and make the last changes we discussed before merging.

wileyj commented on manage.sh:
```diff
@@ -105,6 +105,8 @@ DEFAULT_SERVICES=(
 	stacks-blockchain
 	stacks-blockchain-api
 	postgres
+# Uncomment the following line to export also bitcoin-core logs for troubleshooting
```
not relevant, should be removed - bitcoin is out of scope of the main services this repo *requires* to run, only stacks-blockchain, stacks-blockchain-api, postgres are required services.

criadoperez replied:
Ok, removed comment.

wileyj commented on manage.sh:
```diff
@@ -432,14 +434,72 @@ events_file_env(){
+    fi
+	CONFIG_TOML="${SCRIPTPATH}/conf/${NETWORK}/Config.toml"
+	BTC_CONF="${SCRIPTPATH}/conf/${NETWORK}/bitcoin.conf"
+	${VERBOSE} && log "update_configs BTC_HOST: ${BTC_HOST}"
```
Probably don't need to debug log these variable values

criadoperez replied:
This part isn't my code, it was probably added just to debug.
Ok, commented out.

wileyj commented on manage.sh:
```diff
@@ -432,14 +434,72 @@ events_file_env(){
 # If bitcoin flag is detected when I'm starting the node then
 # the stacks blockchain needs to use a different Config.toml file
 # so it uses the local bitcoin node instead of the remote one
 setup_bitcoin() {
```
outside of printing a verbose log item, what purpose does this function have?
unless it's being used for something, it should be removed

criadoperez replied:
None, it was scheduled to be removed. 
Removed in next commit.

wileyj commented on manage.sh:
```diff
@@ -544,10 +610,11 @@ docker_up() {
-	[[ ! -f "${SCRIPTPATH}/conf/${NETWORK}/Config.toml" ]] && cp "${SCRIPTPATH}/conf/${NETWORK}/Config.toml.sample" "${SCRIPTPATH}/conf/${NETWORK}/Config.toml"
-	[[ ! -f "${SCRIPTPATH}/conf/${NETWORK}/Config-with-bitcoin-flag.toml" ]] && cp "${SCRIPTPATH}/conf/${NETWORK}/Config-with-bitcoin-flag.toml.sample" "${SCRIPTPATH}/conf/${NETWORK}/Config-with-bitcoin-flag.toml"
-	[[ ! -f "${SCRIPTPATH}/conf/${NETWORK}/bitcoin.conf" ]] && cp "${SCRIPTPATH}/conf/${NETWORK}/bitcoin.conf.sample" "${SCRIPTPATH}/conf/${NETWORK}/bitcoin.conf"
+	# [[ ! -f "${SCRIPTPATH}/conf/${NETWORK}/Config.toml" ]] && cp "${SCRIPTPATH}/conf/${NETWORK}/Config.toml.sample" "${SCRIPTPATH}/conf/${NETWORK}/Config.toml"
```
Not needed, the cp commands are changed above

criadoperez replied:
Correct.

wileyj commented on sample.env:
```diff
@@ -58,3 +58,17 @@ POSTGRES_VERSION=14
+
+## Mainnet Defaults
+BTC_HOST=bitcoin.mainnet.stacks.org
+BTC_RPC_USER=stacks-mainnet
```
I'd keep this simple - the rpc user/pass should be kept simple by default (and can be re-used for default setup). 
i'd argue using the existing `stacks:foundation` rpc auth would be fine

criadoperez replied:
I agree. That's how I had it initially! Changed it back.
